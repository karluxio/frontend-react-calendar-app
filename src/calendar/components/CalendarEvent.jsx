export const CalendarEvent = ({ event }) => {

  const { title, notes, start, end, user } = event

  return (
    <>
      <strong>{title}</strong>
      <span> - {user.name}</span>
    </>
  )
}