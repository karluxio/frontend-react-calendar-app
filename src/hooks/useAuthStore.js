import { useDispatch, useSelector } from "react-redux"
import calendarApi from "../api/calendarApi"
import { clearErrorMessage, onChecking, onLogin, onLogout } from "../store"
import { useNavigate } from "react-router-dom"

export const useAuthStore = () => {

  const dispatch = useDispatch()
  const { status, user, errorMessage } = useSelector(state => state.auth)

  const navigate = useNavigate()
  const startLogin = async ({ email, password }) => {

    dispatch(onChecking())

    try {

      const { data } = await calendarApi.post('/auth', { email, password })
      const { token, uid, name } = data
      localStorage.setItem('token', token)
      localStorage.setItem('token-init-date', new Date().getTime())
      dispatch(onLogin({ uid, name }))

    } catch (error) {
      dispatch(onLogout('invalid credentials'))
      setTimeout(() => {
        dispatch(clearErrorMessage())
      }, 10)
    }
  }

  const startRegister = async ({ email, password, name }) => {

    dispatch(onChecking())

    try {

      const { data } = await calendarApi.post('/auth/new', { email, password, name })
      localStorage.setItem('token', data.token)
      localStorage.setItem('token-init-date', new Date().getTime())
      console.log({ uid: data.uid, name: data.name });
      dispatch(onLogin({ uid: data.uid, name: data.name }))

    } catch (error) {
      dispatch(onLogout(error.response.data?.msg || '--'))
      setTimeout(() => {
        dispatch(clearErrorMessage())
      }, 10);
    }
  }

  const checkAuthToken = async () => {
    const token = localStorage.getItem('token')
    if (!token) return dispatch(onLogout())

    try {

      const { uid, name, token } = await calendarApi.get('/auth/renew')
      localStorage.setItem('token', token)
      localStorage.setItem('token-init-date', new Date().getTime())
      dispatch(onLogin({ uid, name }))

    } catch (error) {
      localStorage.clear()
      console.log({ error })
      dispatch(onLogout())
    }
  }

  const startLogout = () => {
    localStorage.clear()
    dispatch(onLogout())
    dispatch(clearErrorMessage())
    navigate('/auth/login')
  }


  return {
    //* Properties
    errorMessage,
    status,
    user,

    //* Methods
    checkAuthToken,
    startLogin,
    startLogout,
    startRegister
  }
}
