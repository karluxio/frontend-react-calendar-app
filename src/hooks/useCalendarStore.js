import { useDispatch, useSelector } from "react-redux"
import { onAddNewEvent, onSetActiveEvent, onUpdateEvent } from "../store"

export const useCalendarStore = () => {

  const { events, activeEvent } = useSelector(state => state.calendar)
  const dispatch = useDispatch()

  const setActiveEvent = (calendarEvent) => {
    dispatch(onSetActiveEvent(calendarEvent))
  }

  const startSavingEvent = (calendarEvent) => {
    if (calendarEvent._id) {
      dispatch(onUpdateEvent({ ...calendarEvent }))
    } else {
      dispatch(onAddNewEvent({ ...calendarEvent, _id: new Date().getTime() }))
    }
  }


  return {
    //* Properties
    activeEvent,
    events,

    //* Methods
    setActiveEvent,
    startSavingEvent
  }
}