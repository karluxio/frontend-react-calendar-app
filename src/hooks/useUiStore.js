import { useDispatch, useSelector } from "react-redux";
import { onCloseDateModal, onOpenDateModal } from "../store";

export const useUiStore = () => {
  const { isDateModalOpen } = useSelector((state) => state.ui);
  const dispatch = useDispatch();

  const openDateModal = () => {
    dispatch(onOpenDateModal());
  };

  const closeDateModal = () => {
    dispatch(onCloseDateModal());
  };

  const toggleDateModal = () => {
    isDateModalOpen ? closeDateModal() : openDateModal();
  };

  return {
    // properties
    isDateModalOpen,

    // functions
    openDateModal,
    closeDateModal,
    toggleDateModal,
  };
};
