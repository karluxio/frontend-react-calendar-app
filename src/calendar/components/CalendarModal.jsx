import { useState, useMemo, useEffect } from "react";

import { addHours, differenceInSeconds } from "date-fns";
import Modal from "react-modal";
import DatePicker, { registerLocale } from "react-datepicker";
import "react-datepicker/dist/react-datepicker.css";
import es from "date-fns/locale/es";
import Swal from "sweetalert2";
import "sweetalert2/dist/sweetalert2.min.css";

import { useCalendarStore, useUiStore } from "../../hooks";
import { getEnvVariables } from "../../helpers";

registerLocale("es", es);

const customStyles = {
  content: {
    top: "50%",
    left: "50%",
    right: "auto",
    bottom: "auto",
    marginRight: "-50%",
    transform: "translate(-50%, -50%)",
  },
};

// to avoid errors in testing environment
if (getEnvVariables().VITE_MODE !== "test") {
  Modal.setAppElement("#root");
}

export const CalendarModal = () => {
  const { isDateModalOpen, closeDateModal } = useUiStore();
  const { activeEvent, startSavingEvent, setActiveEvent } = useCalendarStore();

  const [formSubmitted, setFormSubmitted] = useState(false);

  const [formValues, setFormValues] = useState({
    title: "some title",
    notes: "some notes",
    start: new Date(),
    end: addHours(new Date(), 2),
  });

  const titleClass = useMemo(() => {
    if (!formSubmitted) return "";
    // return (formValues.title.length > 0) ? "is-valid" : "is-invalid"
    return formValues.title.length > 0 ? "" : "is-invalid";
  }, [formValues, formSubmitted]);

  useEffect(() => {
    if (activeEvent) setFormValues({ ...activeEvent });
  }, [activeEvent]);

  const onInputChange = ({ target: { name, value } }) => {
    setFormValues({
      ...formValues,
      [name]: value,
    });
  };

  const onDateChange = (date, changing) => {
    setFormValues({
      ...formValues,
      [changing]: date,
    });
  };

  const onSubmit = (event) => {
    event.preventDefault();
    setFormSubmitted(true);

    const difference = differenceInSeconds(formValues.end, formValues.start);

    if (isNaN(difference) || difference <= 0) {
      Swal("invalid dates", "check dates", "error").catch((e) => e);
      return;
    }

    if (formValues.title.length <= 0) return;

    console.log({ formValues });

    startSavingEvent(formValues);
    closeDateModal();
    setFormSubmitted(false);
  };

  const onCloseModal = () => {
    console.log("close modal");
    closeDateModal();
    setActiveEvent(null);
  };

  return (
    <Modal
      isOpen={isDateModalOpen}
      onRequestClose={onCloseModal}
      style={customStyles}
      contentLabel="Example Modal"
      className="modal"
      overlayClassName="modal-fondo"
      closeTimeoutMS={200}
    >
      <h1> Nuevo evento </h1>
      <hr />
      <form className="container" onSubmit={onSubmit}>
        <div className="form-group mb-2">
          <label>Fecha y hora inicio</label>
          <DatePicker
            locale="es"
            selected={formValues.start}
            onChange={(date) => onDateChange(date, "start")}
            className="form-control"
            dateFormat="Pp"
            showTimeSelect
            timeCaption="Hora"
          />
        </div>

        <div className="form-group mb-2">
          <label>Fecha y hora fin</label>
          <DatePicker
            locale="es"
            minDate={formValues.start}
            selected={formValues.end}
            onChange={(date) => onDateChange(date, "end")}
            className="form-control"
            dateFormat="Pp"
            showTimeSelect
            timeCaption="Hora"
          />
        </div>

        <hr />
        <div className="form-group mb-2">
          <label>Titulo y notas</label>
          <input
            type="text"
            className={`form-control ${titleClass}`}
            placeholder="Título del evento"
            name="title"
            value={formValues.title}
            onChange={onInputChange}
            autoComplete="off"
          />
          <small id="emailHelp" className="form-text text-muted">
            Una descripción corta
          </small>
        </div>

        <div className="form-group mb-2">
          <textarea
            type="text"
            className="form-control"
            placeholder="Notas"
            rows="5"
            name="notes"
            value={formValues.notes}
            onChange={onInputChange}
          ></textarea>
          <small id="emailHelp" className="form-text text-muted">
            Información adicional
          </small>
        </div>

        <button type="submit" className="btn btn-outline-primary btn-block">
          <i className="far fa-save"></i>
          <span> Guardar</span>
        </button>
      </form>
    </Modal>
  );
};
