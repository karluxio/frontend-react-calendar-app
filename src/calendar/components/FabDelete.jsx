import { useDispatch } from "react-redux"
import { useCalendarStore } from "../../hooks"

export const FabDelete = () => {

  const dispatch = useDispatch()
  const { startDeletingEvent, hasEventSelected } = useCalendarStore()

  const onClick = () => {
    dispatch(startDeletingEvent())
  }

  return (
    <button
      className="btn btn-danger fab-delete"
      style={{ display: hasEventSelected ? '' : 'none' }}
      onClick={onClick}
    >
      <i className="fas fa-trash-alt"></i>
    </button>
  )
}