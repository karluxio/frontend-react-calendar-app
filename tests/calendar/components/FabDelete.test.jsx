import { render, screen, fireEvent } from "@testing-library/react";
import { FabDelete } from "../../../src/calendar/components/FabDelete";
import { useCalendarStore } from "../../../src/hooks/useCalendarStore";

jest.mock("../../../src/hooks/useCalendarStore");

const mockStartDeletingEvent = jest.fn();
beforeEach(() => jest.clearAllMocks());
beforeEach(() => jest.clearAllTimers());

describe("<FabDelete />", () => {
  it("should render", () => {
    useCalendarStore.mockReturnValue({
      hasEventSelected: false,
    });

    render(<FabDelete />);

    const btn = screen.getByLabelText("btn-delete");
    // console.log(btn.classList.toString());
    expect(btn.classList).toContain("btn");
    expect(btn.classList).toContain("btn-danger");
    expect(btn.classList).toContain("fab-delete");
    expect(btn.style.display).toBe("none");

    expect(btn.style.display).toBe("none");

    // screen.debug();
  });

  it("should render and show delete button when hasEventSelected=true", () => {
    useCalendarStore.mockReturnValue({
      hasEventSelected: true,
    });

    render(<FabDelete />);

    const btn = screen.getByLabelText("btn-delete");

    expect(btn.style.display).toBe("");
  });

  it("should call startDeletingEvent when btn-delete is clicked", () => {
    useCalendarStore.mockReturnValue({
      hasEventSelected: true,
      startDeletingEvent: mockStartDeletingEvent,
    });

    render(<FabDelete />);

    const btn = screen.getByLabelText("btn-delete");
    fireEvent.click(btn);

    expect(mockStartDeletingEvent).toHaveBeenCalled();
  });
});
