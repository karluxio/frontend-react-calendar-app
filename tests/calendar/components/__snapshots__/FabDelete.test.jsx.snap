// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FabDelete /> should render 1`] = `
<div>
  <button
    class="btn btn-danger fab-delete"
    style="display: none;"
  >
    <i
      class="fas fa-trash-alt"
    />
  </button>
</div>
`;
